File: engine/fen.go
package engine

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

const (
	fenPartsLength          = 6
	fenNoneSymbol           = "-"
	fenSeparator            = " "
	fenPositionSeparator    = "/"
	startingWhiteKingSquare = E1
	startingBlackKingSquare = E8
)

var (
	reIsDigit        = regexp.MustCompile(`^[0-9]$`)
	reEnpSquare      = regexp.MustCompile(`^(-|[abcdefgh][36])$`)
	reCastleRights   = regexp.MustCompile(`(?m)^(-|\bK?Q?k?q?)$`)
	reTurnColor      = regexp.MustCompile(`^(w|b)$`)
	reFenPieceSymbol = regexp.MustCompile(`^[prnbqkPRNBQK]$`)
)

type fenToken struct {
	halfMoveClock uint8
	fullMoveClock uint16
	enpSquare     Square
	castleRights  CastleRights
	turnColor     Color
	position      string
}

// validateFenMetadataParts validates the the parts after the position string token which are
// half and full move clocks, current turn color, en-passant square and castle rights
func validateFenMetadataParts(fen string, opts validateFenOps) (fenToken, error) {
	var emptyRet fenToken

	// tokens length must be 6 after splitting the fen by a single space delimiter
	tokens := strings.Split(fen, fenSeparator)
	if len(tokens) != fenPartsLength {
		return emptyRet, fmt.Errorf("invalid FEN: length must be exactly 6 after splitting by a single space delimiter")
	}

	var (
		fullMoveClockToken = tokens[5]
		halfMoveClockToken = tokens[4]
		enpSquareToken     = tokens[3]
		castleRightsToken  = tokens[2]
		turnColorToken     = tokens[1]
		positionToken      = tokens[0]
	)

	// turn color must be either `w` | `b`
	if !reTurnColor.MatchString(turnColorToken) {
		return emptyRet, fmt.Errorf("invalid FEN: invalid active turn color")
	}

	turn := White
	if turnColorToken == Black.String() {
		turn = Black
	}

	// full move clock must be a number >= 1
	fullMoveClock, err := strconv.ParseUint(fullMoveClockToken, 10, 16)
	if err != nil || fullMoveClock == 0 {
		return emptyRet, fmt.Errorf("invalid FEN: full move clock must be a number >= 1")
	}

	// half move clock must be a number >= 0
	halfMoveClock, err := strconv.ParseUint(halfMoveClockToken, 10, 8)
	if err != nil {
		return emptyRet, fmt.Errorf("invalid FEN: half move clock must be a number >= 0")
	}

	var n uint64
	if turn.IsBlack() {
		n = 1
	}

	// half move clock must be within the limit
	if !(halfMoveClock <= ((fullMoveClock-1)*2)+n) {
		return emptyRet, fmt.Errorf("invalid FEN: half move clock must be whithin the valid limit")
	}

	// in case of an en-passant square, the half move clock must be equal to 0
	if enpSquareToken != fenNoneSymbol && halfMoveClock != 0 {
		return emptyRet, fmt.Errorf("invalid FEN: half move clock must be 0 if en-passant square exists")
	}

	// en-passant square must be a valid square or `-` if empty
	if !reEnpSquare.MatchString(enpSquareToken) {
		return emptyRet, fmt.Errorf("invalid FEN: en-passant target square is invalid")
	}

	enpSquare := SquareNone
	if enpSquareToken != fenNoneSymbol {
		enpSquare = coordToSquare[enpSquareToken]
	}

	if enpSquare != SquareNone {
		if (turn.IsWhite() && enpSquare.Rank() == Rank3) || (turn.IsBlack() && enpSquare.Rank() == Rank6) {
			return emptyRet, fmt.Errorf("invalid FEN: en-passant target square coordinate is invalid")
		}
	}

	// castle rights string must be of valid fen castle string format
	if !reCastleRights.MatchString(castleRightsToken) {
		return emptyRet, fmt.Errorf("invalid FEN: invalid castling rights string")
	}

	var cr CastleRights
	if castleRightsToken != fenNoneSymbol {
		if strings.Contains(castleRightsToken, "K") {
			cr |= WhiteKingSideCastle
		}
		if strings.Contains(castleRightsToken, "Q") {
			cr |= WhiteQueenSideCastle
		}
		if strings.Contains(castleRightsToken, "k") {
			cr |= BlackKingSideCastle
		}
		if strings.Contains(castleRightsToken, "q") {
			cr |= BlackQueenSideCastle
		}
	}

	return fenToken{
		halfMoveClock: uint8(halfMoveClock),
		fullMoveClock: uint16(fullMoveClock),
		enpSquare:     enpSquare,
		castleRights:  cr,
		turnColor:     turn,
		position:      positionToken,
	}, nil
}

// validatePositionPart validates squares and pieces
func validatePositionPart(ft fenToken, opts validateFenOps) (map[Square]Piece, error) {
	ranks := strings.Split(ft.position, fenPositionSeparator)
	if len(ranks) != boardSize {
		return nil, fmt.Errorf("invalid FEN: it does not contain 8 ranks delimited by %q character", fenPositionSeparator)
	}

	piecesCount := make(map[Piece]uint8, 0)
	squares := make(map[Square]Piece, boardTotalSquares)

	for r := 0; r < len(ranks); r++ {
		var sumSquaresInRank uint8
		var previousWasNumber bool

		for f := 0; f < len(ranks[r]); f++ {
			if reIsDigit.MatchString(string(ranks[r][f])) {
				if previousWasNumber {
					return nil, fmt.Errorf("invalid FEN: position string is invalid, it has consecutive numbers")
				}

				n, err := strconv.ParseUint(string(ranks[r][f]), 10, 8)
				if err != nil {
					return nil, fmt.Errorf("invalid FEN: failed to parse row number")
				}

				sumSquaresInRank += uint8(n)
				previousWasNumber = true
			} else {
				piece, err := NewPieceFromFenSymbol(string(ranks[r][f]))
				if err != nil {
					return nil, fmt.Errorf("invalid FEN: position string contains invalid piece symbol")
				}

				sq := Square((7-r)*8 + int(sumSquaresInRank))
				squares[sq] = piece

				piecesCount[piece]++
				sumSquaresInRank++
				previousWasNumber = false
			}
		}

		if sumSquaresInRank != boardSize {
			return nil, fmt.Errorf("invalid FEN: position string is invalid, too many squares in rank")
		}
	}

	if piecesCount[WhiteKing] == 0 {
		return nil, fmt.Errorf("invalid FEN: position is missing white king")
	}
	if piecesCount[BlackKing] == 0 {
		return nil, fmt.Errorf("invalid FEN: position is missing black king")
	}
	if c := piecesCount[WhiteKing]; c > 1 {
		return nil, fmt.Errorf("invalid FEN: position is having too many white kings (%d)", c)
	}
	if c := piecesCount[BlackKing]; c > 1 {
		return nil, fmt.Errorf("invalid FEN: position is having too many black kings (%d)", c)
	}

	for _, char := range ranks[0] {
		if string(char) == WhitePawn.String() {
			return nil, fmt.Errorf("invalid FEN: white pawn is on 8th rank")
		}
	}
	for _, char := range ranks[7] {
		if string(char) == BlackPawn.String() {
			return nil, fmt.Errorf("invalid FEN: black pawn is on 1st rank")
		}
	}

	return squares, nil
}

type validateFenOps struct {
	strict bool
}

type positionMeta struct {
	fenToken
	squares map[Square]Piece
}

// validateFEN validates the fen string
func validateFEN(fen string, opts validateFenOps) (*positionMeta, error) {
	tkn, err := validateFenMetadataParts(fen, opts)
	if err != nil {
		return nil, err
	}

	squares, err := validatePositionPart(tkn, opts)
	if err != nil {
		return nil, err
	}

	meta := positionMeta{
		fenToken: tkn,
		squares:  squares,
	}

	return &meta, nil
}

File: web/src/lib/board/fen.ts
import {
	BLACK,
	BLACK_KING,
	BLACK_PAWN,
	BLACK_QUEEN,
	BOARD_SIZE,
	FEN_EMPTY_POSITION,
	WHITE,
	WHITE_KING,
	WHITE_PAWN,
	WHITE_QUEEN,
} from './common';
import { fenToBoard } from './helpers';
import { Piece } from './piece';
import { Square } from './square';
import { CastleRights, type Color, type Coordinate, type PieceSymbol } from './types';

type FenToken = {
	halfMoveClock: number;
	fullMoveClock: number;
	enpSquare: Square | null;
	castleRights: CastleRights;
	turnColor: Color;
	position: string;
};

const FEN_PARTS_LENGTH = 6;
const FEN_NONE_SYMBOL = '-';
const FEN_SEPARATOR = ' ';
const FEN_POSITION_SEPARATOR = '/';

const RE_IS_DIGIT = /^[0-9]$/;
const RE_ENP_SQUARE = /^(-|[abcdefgh][36])$/;
const RE_CASTLE_RIGHTS = /^(-|\bK?Q?k?q?)$/m;
const RE_TURN_COLOR = /^(w|b)$/;
const RE_FEN_PIECE_SYMBOL = /^[prnbqkPRNBQK]$/;

export function validateFenMetadataParts(fen: string): FenToken {
	// tokens length must be 6 after splitting the fen by a single space delimiter
	const tokens = fen.split(FEN_SEPARATOR);
	if (tokens.length !== FEN_PARTS_LENGTH) {
		throw new Error('invalid FEN: length must be exactly 6 after splitting by a single space delimiter');
	}

	const [positionToken, turnColorToken, castleRightsToken, enpSquareToken, halfMoveClockToken, fullMoveClockToken] =
		tokens;

	// turn color must be either `w` | `b`
	if (!RE_TURN_COLOR.test(turnColorToken)) {
		throw new Error('invalid FEN: invalid active turn color');
	}

	const turn = turnColorToken === 'b' ? BLACK : WHITE;

	// full move clock must be a number >= 1
	const fullMoveClock = Number.parseInt(fullMoveClockToken);
	if (Number.isNaN(fullMoveClock) || fullMoveClock < 1) {
		throw new Error('invalid FEN: full move clock must be a number >= 1');
	}

	// half move clock must be a number >= 0
	const halfMoveClock = Number.parseInt(halfMoveClockToken);
	if (Number.isNaN(halfMoveClock) || halfMoveClock < 0) {
		throw new Error('invalid FEN: half move clock must be a number >= 0');
	}

	const n = turn === BLACK ? 1 : 0;

	// half move clock must be within the limit
	if (!(halfMoveClock <= (fullMoveClock - 1) * 2 + n)) {
		throw new Error('invalid FEN: half move clock must be whithin the valid limit');
	}

	// in case of an en-passant square, the half move clock must be equal to 0
	if (enpSquareToken !== FEN_NONE_SYMBOL && halfMoveClock !== 0) {
		throw new Error('invalid FEN: half move clock must be 0 if en-passant square exists');
	}

	// en-passant square must be a valid square or `-` if empty
	if (!RE_ENP_SQUARE.test(enpSquareToken)) {
		throw new Error('invalid FEN: en-passant target square is invalid');
	}

	let enpSquare: Square | null = null;
	if (enpSquareToken !== FEN_NONE_SYMBOL) {
		enpSquare = Square.fromCoord(enpSquareToken as Coordinate);
	}

	if (enpSquare !== null) {
		if ((turn === WHITE && enpSquare.rank === 3) || (turn === BLACK && enpSquare.rank === 6)) {
			throw new Error('invalid FEN: en-passant target square coordinate is invalid');
		}
	}

	// castle rights string must be of valid fen castle string format
	if (!RE_CASTLE_RIGHTS.test(castleRightsToken)) {
		throw new Error('invalid FEN: invalid castling rights string');
	}

	let cr: CastleRights = CastleRights.None;

	if (castleRightsToken.includes(WHITE_KING)) {
		cr |= CastleRights.WhiteKingSide;
	}
	if (castleRightsToken.includes(WHITE_QUEEN)) {
		cr |= CastleRights.WhiteQueenSide;
	}
	if (castleRightsToken.includes(BLACK_KING)) {
		cr |= CastleRights.BlackKingSide;
	}
	if (castleRightsToken.includes(BLACK_QUEEN)) {
		cr |= CastleRights.BlackQueenSide;
	}

	return {
		halfMoveClock,
		fullMoveClock,
		enpSquare,
		castleRights: cr,
		turnColor: turn,
		position: positionToken,
	};
}

function validatePositionPart(ft: FenToken): void {
	const ranks = ft.position.split(FEN_POSITION_SEPARATOR);
	if (ranks.length !== BOARD_SIZE) {
		throw new Error(`invalid FEN: it does not contain 8 ranks delimited by ${FEN_POSITION_SEPARATOR} character`);
	}

	const piecesCount: Record<string, number> = {};

	for (let r = 0; r < ranks.length; r++) {
		let sumSquaresInRank = 0;
		let previousWasNumber = false;

		for (let f = 0; f < ranks[r].length; f++) {
			if (RE_IS_DIGIT.test(ranks[r][f])) {
				if (previousWasNumber) {
					throw new Error('invalid FEN: position string is invalid, it has consecutive numbers');
				}

				const n = Number.parseInt(ranks[r][f]);
				if (Number.isNaN(n) || n < 0) {
					throw new Error('invalid FEN: failed to parse row number');
				}

				sumSquaresInRank += n;
				previousWasNumber = true;
			} else {
				if (!RE_FEN_PIECE_SYMBOL.test(ranks[r][f])) {
					throw new Error('invalid FEN: invalid piece symbol');
				}
				const piece = Piece.fromPieceFenSymbol(ranks[r][f] as PieceSymbol);
				piecesCount[piece.toFenSymbol()]++;
				sumSquaresInRank++;
				previousWasNumber = false;
			}
		}

		if (sumSquaresInRank != BOARD_SIZE) {
			throw new Error('invalid FEN: position string is invalid, too many squares in rank');
		}
	}

	if (piecesCount[WHITE_KING] === 0) {
		throw new Error('invalid FEN: position is missing white king');
	}
	if (piecesCount[BLACK_KING] === 0) {
		throw new Error('invalid FEN: position is missing black king');
	}

	const wkc = piecesCount[WHITE_KING];
	if (wkc > 1) {
		throw new Error(`invalid FEN: position is having too many white kings ${wkc}`);
	}

	const bkc = piecesCount[BLACK_KING];
	if (bkc > 1) {
		throw new Error(`invalid FEN: position is having too many black kings ${bkc}`);
	}

	for (const char of ranks[0]) {
		if (char === WHITE_PAWN) {
			throw new Error('invalid FEN: white pawn is on 8th rank');
		}
	}

	for (const char of ranks[7]) {
		if (char === BLACK_PAWN) {
			throw new Error('invalid FEN: black pawn is on 1st rank');
		}
	}
}

export type PositionMeta = {
	fenToken: FenToken;
	squares: Square[];
};

export function validateFen(fen: string): PositionMeta {
	if (fen === FEN_EMPTY_POSITION) {
		const squares = fenToBoard(fen);

		return {
			fenToken: {
				castleRights: CastleRights.None,
				enpSquare: null,
				fullMoveClock: 1,
				halfMoveClock: 0,
				position: fen.split(' ')[0],
				turnColor: WHITE,
			},
			squares,
		};
	}

	const fenToken = validateFenMetadataParts(fen);
	validatePositionPart(fenToken);

	const squares = fenToBoard(fenToken.position);

	return {
		fenToken,
		squares,
	};
}
